### ISSUE TITLE
Install issues

### ISSUE THREAD

**Issue by pfitzseb** [prebuild/prebuild-install#77]
Hey everyone,

we've gotten reports about failed installs for [`atom-julia-client`](https://github.com/JunoLab/atom-julia-client), which uses [`node-pty-prebuilt`](https://github.com/daviwil/node-pty-prebuilt), which in turn makes use of this package.

I'm not sure if this is actually the right place for this issue -- feel free to close this if not.

[Here](https://github.com/JunoLab/atom-julia-client/issues/440)'s the issue at our repo. The failure on Windows is probably some connection issue:
```
prebuild-install info unpacking @ C:\Users\Farhan\.atom\.apm\_prebuilds\https-github.com-daviwil-node-pty-prebuilt-releases-download-v0.7.3-node-pty-prebuilt-v0.7.3-electron-v53-win32-x64.tar.gz
prebuild-install WARN install unexpected end of file
```
even though it apparently was not resolved even when retrying. The Linux issue on the other hand seems consistent:
```
prebuild-install WARN install ENAMETOOLONG: name too long, open '/home/user/.atom/.apm/_prebuilds/https-github.com-daviwil-node-pty-prebuilt-releases-download-v0.7.3-node-pty-prebuilt-v0.7.3-electron-v54-linux-x64.tar.gz.10302-1f746b13a169.tmp'
```
AFAIK no OS should have such a low file/path length limit though (the above is ~185 chars for the path and ~145 chars for the filename). I'm not sure how to debug this further though (can't reproduce either issue locally and the debug logs are apparently not available) or what a possible solution might be -- possibly making sure that the file name is much shorter or something....

Thanks!

**Comment by ralphtheninja** [comment]
@pfitzseb What's the command to reproduce this?

**Comment by pfitzseb** [comment]
In theory you should be able to reproduce with `npm install node-pty-prebuilt` or something like that.
However, I haven't been able to reproduce this in any way -- or seems like a very spurious issue.

**Comment by ralphtheninja** [comment]
I tried using node `v9.11.1` (no prebuilts for node 10 have been made) and get this output

`npm i --verbose`:

```
prebuild-install info begin Prebuild-install version 2.5.3
prebuild-install info looking for local prebuild @ prebuilds/node-pty-prebuilt-v0.7.3-node-v59-linux-x64.tar.gz
prebuild-install info looking for cached prebuild @ /home/lms/.npm/_prebuilds/https-github.com-daviwil-node-pty-prebuilt-releases-download-v0.7.3-node-pty-prebuilt-v0.7.3-node-v59-linux-x64.tar.gz
prebuild-install info found cached prebuild                            
prebuild-install info unpacking @ /home/lms/.npm/_prebuilds/https-github.com-daviwil-node-pty-prebuilt-releases-download-v0.7.3-node-pty-prebuilt-v0.7.3-node-v59-linux-x64.tar.gz
prebuild-install info unpack resolved to /home/lms/tmp/node-pty-test/node_modules/node-pty-prebuilt/build/Release/pty.node
prebuild-install info unpack required /home/lms/tmp/node-pty-test/node_modules/node-pty-prebuilt/build/Release/pty.node successfully
prebuild-install info install Successfully installed prebuilt binary!             
```

So it seems to be working fine (on linux at least).

@pfitzseb Have you tried reproducing on windows?

**Comment by ralphtheninja** [comment]
Just some background; `prebuild-install` caches all prebuilts in `~/.npm/_prebuilds` and takes the URL and converts it to a proper file name, so 

`https://github.com/daviwil/node-pty-prebuilt/releases/download/v0.7.3/node-pty-prebuilt-v0.7.3-node-v59-linux-x64.tar.gz`

becomes

`https-github.com-daviwil-node-pty-prebuilt-releases-download-v0.7.3-node-pty-prebuilt-v0.7.3-node-v59-linux-x64.tar.gz`

In theory we could try to be a little smarter her. Currently we're just replacing invalid characters with `'-'`.

**Comment by mathiask88** [comment]
I tried to reproduce this as well, but was not able to. I think more interesting is the OS/Filesystem. At least it is no `prebuild-install` issue, but as @ralphtheninja said maybe we could be a bit smarter in naming IF long paths are the reason.

**Comment by pfitzseb** [comment]
So apparently [there's a 143 byte filename limit](https://askubuntu.com/questions/728465/how-can-i-increase-ubuntus-143-byte-file-name-limit-for-encrypted-directories/728525) on encrypted Ubuntu filesystems.

Would you accept a PR for making the filename a (MD5?) hash of the URL or something similar that ensures a fixed size filename (much) shorter than 140 chars?

**Comment by ralphtheninja** [comment]
> Would you accept a PR for making the filename a (MD5?) hash of the URL or something similar that ensures a fixed size filename (much) shorter than 140 chars?

Def. If you want to hack something up for discussion.

**Comment by vweevers** [comment]
-1 on hashing as it would make the files hard to identify for humans. How about something like:

`/home/user/.npm/_prebuilds/daviwil-node-pty-v0.7.3-node-v59-linux-x64.tar.gz` (76 chars)

Not sure how'd it work with [custom binaries](https://github.com/prebuild/prebuild-install#custom-binaries) though.

**Comment by ralphtheninja** [comment]
@vweevers Agree on the downside of hashing. The file name in itself should be good enough imo.

**Comment by ralphtheninja** [comment]
Just a side note. When switching to another caching algorithm the current cache will be stale. Should we do something about that or just leave it be? Does this mean a new major version?

**Comment by mathiask88** [comment]
-1 on hashing as well for the same reasons. What about reducing the file name and append the reduced information to the path instead? So we utilize the PATH_MAX which seems to be always higher or at least equal than NAME_MAX.

**Comment by vweevers** [comment]
> What about reducing the file name and append the reduced information to the path instead?

That could work, if we really need that reduced information.

**Comment by vweevers** [comment]
If you switch to a mirror, should the filename (or path) change?

**Comment by pfitzseb** [comment]
**Current**:
`~/.npm/_prebuilds/https-github.com-daviwil-node-pty-prebuilt-releases-download-v0.7.3-node-pty-prebuilt-v0.7.3-node-v59-linux-x64.tar.gz`

- **Option 1** (what @mathiask88 suggests, if I understand correctly):
`~/.npm/_prebuilds/https-github.com/daviwil/node-pty-prebuilt/releases/download/v0.7.3/node-pty-prebuilt-v0.7.3-node-v59-linux-x64.tar.gz`

- **Option 2** (prefixed partial checksum):
`~/.npm/_prebuilds/71a2427262-node-pty-prebuilt-v0.7.3-node-v59-linux-x64.tar.gz` 
where `71a2427262` are the first 10 chars of the download URL checksum.

- **Option 3** (checksum only):
`~/.npm/_prebuilds/71a242726206b5e537d3a82de5e16791.tar.gz`

Option 3 is probably out of the question for the reasons listed above. I kinda like Option2 since it doesn't create so many directories (at the cost of some info though).

**Comment by ralphtheninja** [comment]
Fix released in 4.0.0
